### PR TITLE
feature/18-import

### LINKED ISSUE #18: Import support from command line.
For users that prefer to stay away from the web vault and for scripting bulk operations.

Benefits:
* unencrypted data never has to touch a web browser
* self-hosting the backend requires extra effort
* for people who use a non-standard backend (I'm currently in love with https://github.com/vvondra/bitwarden-serverless but this is also true for whoever uses the ruby implementation) there's a fair amount of extra work to get the vault running.

It would be awesome to support doing this from the CLI, since 90% of the work is already done in jslib.

I intend to take on this work and implement it as follows:

1. Recreate the generic import logic from web in jslib as `importer.service`.
2. wrap that service from the CLI just like export does.
3. open jslib pull request, and once it's closed and released, open the cli changes as another.
4. go back and refactor web to use the new importer service, and submit as a separate pull request.

Please let me know if you have reservations about any of that.

### PR DESCRIPTION
**Pull request by subdavis**
Tested working with core-as-a-service and self-hosted core.

Fixed 

```
(node:14012) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
```

since I was already mucking about also.

Requires bitwarden/jslib#8

fixes #18

### CODE REVIEW

File: src/program.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ import { ListResponse } from './models/response/listResponse';
 import { MessageResponse } from './models/response/messageResponse';
 import { StringResponse } from './models/response/stringResponse';
 import { TemplateResponse } from './models/response/templateResponse';
+import { format } from 'url';
```
Comment (kspearrin) [inline src/program.ts]: unused import?

File: src/utils.ts [hunk 2]
```diff
@@ -9,6 +9,29 @@ import { FolderView } from 'jslib/models/view/folderView';
 import { NodeUtils } from 'jslib/misc/nodeUtils';
 
 export class CliUtils {
+    static readFile(input: string): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
+            let p: string = null;
+            console.log("here");
```
Comment (kspearrin) [inline src/utils.ts]: remove log

File: src/commands/import.command.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        const email = await this.userService.getEmail();
+        const key = await this.cryptoService.makeKey(password, email);
+        const keyHash = await this.cryptoService.hashPassword(password, key);
+        const storedKeyHash = await this.cryptoService.getKeyHash();
+        const importer = await this.importService.getImporter(type);
+        if (importer === null) {
+            return Response.badRequest('Proper importer type required.');
+        }
+        if (storedKeyHash != null && keyHash != null && storedKeyHash === keyHash) {
+            return CliUtils.readFile(path).then(async contents => {
```
Comment (kspearrin) [inline src/commands/import.command.ts]: lint issue here i think. need parens. do `npm run lint`

File: src/commands/import.command.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                let submitResult = await this.importService.submit(importer, contents);
```
Comment (kspearrin) [inline src/commands/import.command.ts]: another lint issue, const

File: src/program.ts [hunk 5]
```diff
@@ -395,6 +397,22 @@ export class Program {
...
+            .description('Import vault data from a file.')
+            .option('-l, --list', 'List valid formats')
+            .on('--help', () => {
+                writeLn('\n Examples:');
+                writeLn('');
+                writeLn('    bw import bitwardencsv ./from/source.csv')
+                writeLn('    bw import keepass2xml keepass_backup.xml myPassword123')
+            })
+            .action(async (format, filepath, password, cmd) => {
+                const command = new ImportCommand(this.main.cryptoService, this.main.userService, this.main.importService);
```
Comment (kspearrin) [inline src/program.ts]: need an `await this.exitIfLocked();`

**Comment by kspearrin** [comment]
also, revert package lock changes please. i will regenerate.

**Comment by subdavis** [comment]
Thanks for the comments.  I'll get these and the fixes for jslib together.

File: src/program.ts (hunk 5 above)
Comment (kspearrin) [reply src/program.ts]: Still need this locked check

**Comment by kspearrin** [comment]
Still missing lock check. Also, please dont squash on updates. I can't see what changed. I will squash on merge.

**Comment by subdavis** [comment]
Added locked check for sensitive operations.  Listing formats is possible without an unlocked account.

**Comment by kspearrin** [comment]
Thanks for the PR. I made a few adjustments. Works great,

**Pull request merged**